20250322-actors-pubsub.md

```diff
@@ -0,0 +1,367 @@
+
+To that end, I propose that Dapr build a Subscription Manager that's semi-compliant with the current CloudEvents
+[subscription specification](https://github.com/cloudevents/spec/blob/main/subscriptions/spec.md)
+
+A given subscription manager is responsible for one or more subscriptions assigned it, though this assignment can be
+optimized to group like sources and subjects to minimize unnecessary scans for inbound events.
+
+The design choice has been previously made to use the Common Expression Language for routing and this proposal does
+not diverge from this decision. Accordingly, as this **DOES** differ from the CloudEvents specification, the example
+below changes the `filter` property to `rules` and updates the signature accordingly and puts the sink in the
+`rule` body accordingly along with an optional default endpoint to which the delivery is made irrespective of any
+other rules.
+
+A subscription, per this specification, is modeled using JSON as follows:
+
+```json
+{
+  // Required: The unique identifier of the subscription in the scope of the subscription manager
+  "id": "[a subscription manager scoped unique string]",
+  // Required: The source to which the subsription is related (in Dapr, the pubsub component name
+  "source": "<pubsub_component_name>",
+  // Required: The topic the subscription is configured for
+  "subject": "<pubsub_topic_name>",
+  // Optional: An array of filter expressions that evaluate to true or false. Delivery should be performed to the 
+  // identified sink only if the CEL expression evaluates to true. Rules are evaluated indepednently of one another.
+  "routes": {
+    "rules": [
+      // If either of the `match` or `sink` properties are provided, both must be
+      {
+        // Required: The CEL expression that must evaluate as true to match on this route 
+        "match": "",
+        // Required: The destination actor to which events MUST be sent if the expression is true 
+        // The last segment, "/<actor_id>" is optional and should only be present for a subscription that is specific to an actor instance
+        "sink": "https://dapr.io/events/<actor_type>/<actor_method>/<actor_id>"
+      },
+      // Optional: The destination actor to which events MUST be sent if the expression is true
+      // The last segment, "/<actor_id>" is optional and should only be present for a subscription that is specific to an actor instance
+      {
+        "default": "https://dapr.io/events/<actor_type>/<actor_method>/<actor_id>"
+      }
+    ]
+  },
+  // Required: The identifier of the delivery protocol, whether HTTP or gRPC
+  "protocol": "HTTP"
+}
+```
+
+Whether a PubSub subscription is implemented via the HTTP or gRPC protocols, it should result in the creation of the above
+subscription so that there is no substantive difference in how one subscription is registered versus another, lessening
+the burden on the runtime to juggle different approaches.
+
+#### Streaming Subscriptions API
```

Comment:
Are we going to implement this in a phased approach? Like start with streaming subscriptions then add to declarative and programmatic? What is the thought process here for scope?

Reply:
I think the result of our chat Monday was that we should look at supporting streaming subscriptions first and only revisit with the remaining options later if there's demand for it. Declarative supports hot reload, but that requires extra-code changes to make and programmatic is fixed at app load (and is finnicky with actors because of the activation requirement), so having an intentional opt-in at the actor instance level is the most straightforward route.

I'll modify the proposal itself to reflect this in the next day or two.